python/cudf_polars/cudf_polars/experimental/parallel.py

```diff
@@ -0,0 +1,120 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+"""Partitioned LogicalPlan nodes."""
+
+from __future__ import annotations
+
+from functools import singledispatch
+from typing import TYPE_CHECKING, Any
+
+from cudf_polars.dsl.expr import NamedExpr
+from cudf_polars.dsl.ir import PartitionInfo
+from cudf_polars.dsl.traversal import traversal
+
+if TYPE_CHECKING:
+    from collections.abc import MutableMapping
+
+    from cudf_polars.containers import DataFrame
+    from cudf_polars.dsl.ir import IR
+    from cudf_polars.dsl.nodebase import Node
+
+
+def get_key_name(node: Node | NamedExpr) -> str:
+    """Generate the key name for a Node."""
+    if isinstance(node, NamedExpr):
+        return f"named-{get_key_name(node.value)}"
+    return f"{type(node).__name__.lower()}-{hash(node)}"
+
+
+@singledispatch
+def lower_ir_node(ir: IR, rec) -> IR:
+    """Rewrite an IR node with proper partitioning."""
+    # Return same node by default
+    return ir
```

Comment:
```suggestion
@singledispatch
def lower_ir_node(ir: IR, rec) -> IR:
    """Rewrite an IR node with proper partitioning."""
    # Return same node by default
    return ir.reconstruct([rec(child) for child in ir.children])
```
Ah. While playing with groupby rewrites in a separate branch, I realized that we need to reconstruct here. @wence- - Does this suggested change make sense?